### ISSUE TITLE
finalizeExit() DoS attack

### ISSUE THREAD

**Issue by laskdaf** [omisego/plasma-mvp#66]
If transfer() fails in finalizeExit(), then the entire function reverts, stopping any future exits from being processes. The PQ essentially cannot pop off the malicious exit. One could easily exploit this by writing a contract that has a payable fall back function: 

`function() payable public {
        revert();
}`

This can be solved by moving the transfer() function to a separate withdraw function that the owner must call. We would map each user's address to their balance, and the withdraw function will refund them. 

- Blockchain @ Berkeley Researcher

**Comment by kfichter** [comment]
Good catch. Also, what happens if the exit queue is sufficiently large that the while loop runs out of gas even with max gas provided? I believe the whole transaction would revert and the exits can no longer be processed.

**Comment by laskdaf** [comment]
Rather than doing it in a loop (having one user brunt the massive gas cost), we can make each finalizeExit() process one exit at a time. This solves the maxing out the gas cost problem and distributes the burden onto more people. Anyone in the queue has the incentive to call finalizeExit() as they want their exit to be processed.

**Comment by kfichter** [comment]
I guess we could also check `msg.gas` and require some minimum gas before each loop, else return

**Comment by laskdaf** [comment]
Yes, that way each user who calls finalizeExit() determines the number of exits one wishes to process. As far as I know, the minimum gas cost for each loop would need to be hardcoded in by looking at the compiled bytecode.
